gossip/src/duplicate_shred_handler.rs

```diff
@@ -63,10 +60,16 @@ pub struct DuplicateShredHandler {
     // We don't want bad guys to inflate the chunk map, so we limit the number of
     // pending proofs from each pubkey to ALLOWED_SLOTS_PER_PUBKEY.
     validator_pending_proof_map: HashMap<Pubkey, HashSet<Slot>>,
-    // remember the last root slot handled, clear anything older than last_root.
-    last_root: Slot,
+    // Cache last root to reduce read lock.
+    last_root: Option<Slot>,
+    // remember the last root slot cleaned, clear anything between last_root and last_root_cleaned.
+    last_root_cleaned: Option<Slot>,
     blockstore: Arc<Blockstore>,
     leader_schedule_cache: Arc<LeaderScheduleCache>,
+    bank_forks: Arc<RwLock<BankForks>>,
+    // Cache information from root bank so we could function correctly without reading roots.
+    cached_staked_nodes: Option<Arc<HashMap<Pubkey, u64>>>,
+    cached_slots_in_epoch: u64,
```

Comment:
any reason not to keep this as a constant like before/

Reply:
I think Carl and Behzad suggested that this should be as long as an epoch?

Reply:
but you have it set to 100, and it's not modified at all right? could just keep it as a `const` and change it to the size of an epoch

Reply:
can we use `Clock::DEFAULT_SLOTS_PER_EPOCH`?

Reply:
actually nvm this is probably better as it works for tests which have custom epoch schedules